File: src/ai/scheduler/src/scheduler_node.cpp
#include "scheduler/scheduler_node.hpp"

/**
 * @brief object constructor
 *
 * @param n NodeHandler
 */
Scheduler::Scheduler() : PerformClient("scheduler", "ai"), side(Side::LEFT), robotPosition() {
	// Retrieve action
	std::string actions_file;
	nh.getParam("actions_directory", actions_file);

	this->side = Side::LEFT;
	this->control_srv = nh.advertiseService("/scheduler/do", &Scheduler::setState, this);

	this->robotPosition_sub = nh.subscribe<can_msgs::Point>("/STM/Position", 10, &Scheduler::setRobotPosition, this);

	// Parse actions file
	try {
		ActionsParser parser(ActionFilePath("main", actions_file));
		this->rootAction = parser.getAction();
	} catch(const char* error) {
		ROS_ERROR_STREAM("Unable to initialize scheduler due to parsing error: " << error);
		setNodeStatus(NodeStatus::NODE_ERROR, 1);
		return;
	}

	std::vector<NodeRequirement> reqs;
	this->getRequired(reqs, this->rootAction);

	this->waitForNodes(reqs, 5);
}

void Scheduler::onWaitingResult(bool success) {
	if (success) {
		setNodeStatus(NodeStatus::NODE_READY);
	} else {
		ROS_ERROR_STREAM("Some actions are missing, unable to start node");
		setNodeStatus(NodeStatus::NODE_ERROR, 2);
		return;
	}
}

bool Scheduler::setState(SetSchedulerState::Request &req, SetSchedulerState::Response &res) {
	// If there is a change in state
	if (this->running != req.running) {
		this->running = req.running;
		
		// apply change
		if (req.running) {
			ROS_INFO("Scheduler actions resumed !");
			this->side = req.side;

			// Resume action
			this->nextAction();
		} else {
			ROS_INFO("Scheduler actions paused...");
			cancelAction();
		}
	}

	return true;
}

void Scheduler::nextAction() {
	if (!this->running) {
		return;
	}

	// If pause cancer has spread into the root of the action tree
	if (this->rootAction->state() == ActionStatus::PAUSED) {
		// We cure it
		this->rootAction->setState(ActionStatus::IDLE);
	}
	
	if (this->rootAction->state() == ActionStatus::IDLE) {
		ActionChoice choice =
			getOptimalNextAtomic(this->rootAction, this->robotPosition);

		ROS_INFO_STREAM("[Current action tree]" << std::endl << *this->rootAction);

		if (choice.action != NULL) {
			// save current action
			this->currentAction = choice.action;

			ROS_INFO_STREAM("Perfoming : " << *this->currentAction);

			// call then onfinished or onpause
			this->performAction(choice.action, this->robotPosition);
		} else {
			ROS_INFO_STREAM("No action to be performed");
			this->rootAction->setState(ActionStatus::DONE);
		}
	} else {
		ROS_INFO_STREAM("Root action is in state " << this->rootAction->state());
	}
}

void Scheduler::onFinished(const actionlib::SimpleClientGoalState& state, const ai_msgs::PerformResultConstPtr& result) {
	currentAction->setState(ActionStatus::DONE);
	nextAction();
}

void Scheduler::onPaused() {
	currentAction->setState(ActionStatus::PAUSED);
	nextAction();
}

void Scheduler::setRobotPosition(const can_msgs::Point::ConstPtr& msg) {
	this->robotPosition.x = msg->pos_x;
	this->robotPosition.y = msg->pos_y;
	this->robotPosition.angle = msg->angle;
}

int main(int argc, char *argv[]) {
	ros::init(argc, argv, "scheduler_node");
	Scheduler node;
	ros::spin();
}

File: src/ai/scheduler/src/scheduler_node.py
#!/usr/bin/python3

import rospy
from rospkg import RosPack

from action_manager import PerformClient, ActionGroup, Action

from interface_msgs.msg import CanData
from ai_msgs.msg import Side, NodeStatus, ActionStatus
from ai_msgs.srv import SetSchedulerState, SetSchedulerStateRequest, SetSchedulerStateResponse
from geometry_msgs.msg import Pose2D

from xml_class_parser import ParsingException

class SchedulerNode(PerformClient):
	def __init__(self):
		super().__init__("scheduler", "ai")

		self.running = False
		self.current_action: Action = None
		self.robot_pos = Pose2D()

		# ROS services / subscribers
		self.control_srv = rospy.Service("/scheduler/do", SetSchedulerState, self.set_state)
		self.robot_pos_sub = rospy.Subscriber("/can_interface/in", CanData, self.on_can_message)

		# Parse actions
		try:
			folder = RosPack().get_path("ai_description") + "/actions/"
			self.root_action = ActionGroup.parse_file(
				folder + "main.xml", context={"folder": folder})
		except ParsingException as e:
			self.set_status(NodeStatus.ERROR, 1)
			raise e

		# Load requirements
		self.save_required(self.root_action)
		self.wait_for_nodes(5)

	def on_waiting_result(self, success: bool):
		if success:
			self.set_status(NodeStatus.READY)
		else:
			rospy.logerr("Some actions are missing, unable to start node")
			self.set_status(NodeStatus.ERROR, 2)

	def set_state(self, request: SetSchedulerStateRequest):
		"""
			Service function to change scheduler. It change state only if
			requested state is different
		"""
		# If there is a change in state
		if self.running != request.running:
			self.running = request.running
			
			# apply change
			if request.running:
				rospy.loginfo("Scheduler actions resumed !")
				
				# Set side
				self.root_action.set_side(request.side)

				# Resume action
				self.next_action()
			else:
				rospy.loginfo("Scheduler actions paused...")
				self.cancel_action()

		return SetSchedulerStateResponse()

	def on_can_message(self, data: CanData):
		"""
			Handle robot position from STM
		"""
		# Right type of data
		if data.type == "robot_position":
			self.robot_pos.x = data.params.get("x", int)
			self.robot_pos.y = data.params.get("y", int)
			self.robot_pos.theta = data.params.get("theta", int)

	def on_action_returns(self, state, result):
		"""
			Current action done
		"""
		self.current_action.state = result.state
		self.next_action()

	def next_action(self):
		if not self.running:
			return
		
		# Following peace of code assert once a group is launched, it's actions
		# are performing until group state is not IDLE
		current_root = self.root_action
		if self.current_action is not None:
			# Compute local root to continue previously selected group until it is done
			current_root = self.current_action

			while current_root.state != ActionStatus.IDLE and \
				current_root.parent is not None:

				current_root = current_root.parent


		# If pause cancer has spread into the root of the action tree
		if self.root_action.state == ActionStatus.PAUSED:
			# We cure it
			self.root_action.state = ActionStatus.IDLE
		
		rospy.loginfo("[Current action tree]\n{}".format(self.root_action))

		# If there is alive actions again
		if self.root_action.state == ActionStatus.IDLE:
			choice: ActionChoice = self.root_action.get_optimal(self.robot_pos)


			if choice.action is not None:
				# save current action
				self.current_action = choice.action

				rospy.loginfo("Perfoming : {}".format(self.current_action))

				# call then onfinished or onpause
				self.perform_action(choice.action, self.robot_pos)
			else:
				rospy.loginfo("No action to be performed")
				self.root_action.state = ActionStatus.DONE
			
		else:
			# Otherwise signal that everything is over :)
			rospy.loginfo("Root action is in state {}, nothing more to be done...".format(self.root_action.state))
		
if __name__ == "__main__":
	rospy.init_node("scheduler_node", log_level=rospy.INFO)
	node = SchedulerNode()

	while not rospy.is_shutdown():
		rospy.spin()